currency/pairs_test.go

```diff
@@ -267,19 +267,21 @@ func TestAdd(t *testing.T) {
 		NewPair(LTC, USD),
 		NewPair(LTC, USDT),
 	}
-
 	// Test adding a new pair to the list of pairs
 	p := NewPair(BTC, USDT)
 	pairs = pairs.Add(p)
 	if !pairs.Contains(p, true) || len(pairs) != 4 {
 		t.Error("TestAdd unexpected result")
 	}
-
 	// Now test adding a pair which already exists
 	pairs = pairs.Add(p)
 	if len(pairs) != 4 {
 		t.Error("TestAdd unexpected result")
 	}
+	pairs = pairs.Add(NewPair(BTC, LTC), NewPair(ETH, USD))
+	if len(pairs) != 6 {
+		t.Error("TestAdd unexpected result")
+	}
```

Comment:
There's a non zero chance of append not allocating a separate slice and you get weird effects and spooky ghosts 👻. Here's a test for it:

```diff
diff --git a/currency/pairs_test.go b/currency/pairs_test.go
index 33d3fa411..1eaa494f6 100644
--- a/currency/pairs_test.go
+++ b/currency/pairs_test.go
@@ -3,6 +3,7 @@ package currency
 import (
        "encoding/json"
        "errors"
+       "fmt"
        "testing"

        "github.com/stretchr/testify/assert"
@@ -282,6 +283,43 @@ func TestAdd(t *testing.T) {
        if len(pairs) != 6 {
                t.Error("TestAdd unexpected result")
        }
+
+       pairsWithExtraBaggage := make(Pairs, 0, len(pairs)+3)
+       pairsWithExtraBaggage = append(pairsWithExtraBaggage, pairs...)
+
+       brain := NewPair(BRAIN, USD)
+       withBrain := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), brain)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, brain, withBrain[len(withBrain)-1])
+
+       badger := NewPair(BADGER, USD)
+       withBadger := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), badger)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, badger, withBadger[len(withBadger)-1])
+
+       babyDoge := NewPair(BABYDOGE, USD)
+       withBabyDoge := pairsWithExtraBaggage.Add(NewPair(BTC, LTC), babyDoge)
+       if len(pairs) != 6 {
+               t.Error("TestAdd unexpected result")
+       }
+
+       assert.Equal(t, babyDoge, withBabyDoge[len(withBabyDoge)-1])
+
+       // check bleed back
+
+       assert.Equal(t, brain, withBrain[len(withBrain)-1])
+       assert.Equal(t, badger, withBadger[len(withBadger)-1])
+
+       fmt.Println(withBadger)
+       fmt.Println(withBrain)
+       fmt.Println(withBabyDoge)
+       fmt.Println(pairsWithExtraBaggage)
 }
```

slices.Clone call before or make(Pairs, 0, len(p) + len(Pairs) then copy original acrost.